### PR TITLE
[SOC-5485] Refused Spaces suggestions should not be displayed again

### PR DESCRIPTION
**Pull request by AtiaZ**
fix description : 
1- add a new SpaceMember Status and a new MembershipType : IGNORED.
2- add createSpaceMembershipForIgnoredStatus() method to change MembershipType and setIgnored() method to change the SpaceMember Status in the SpaceMemberEntity when we ignore a space.
3- add isIgnore() in the interface class and isSpaceIgnored() in the SpaceStorage interface to verify if the given space is ignored by the user or not and then avoid to display it in the list of suggested spaces.

### CODE REVIEW

File: component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/RDBMSSpaceStorageImpl.java [hunk 1]
```diff
@@ -413,6 +413,28 @@ public void renameSpace(Space space, String newDisplayName) throws SpaceStorageE
...
 
   @Override
+  public void updateSpaceMembership(String spaceId, String userId) {
+    SpaceMemberEntity entity = spaceMemberDAO.getSpaceMemberShip(userId, Long.parseLong(spaceId));
+    SpaceEntity spaceEntity = spaceDAO.find(Long.parseLong(spaceId));
+    if (entity == null) {
+      entity = new SpaceMemberEntity();
+      entity.setSpace(spaceEntity);
+      entity.setUserId(userId);
+      entity.setStatus(Status.IGNORED);
```
Comment (thomasdelhomenie) [inline component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/RDBMSSpaceStorageImpl.java]: The IGNORED status is always used in this method. Does not seem normal for a method called updateSpaceMembership.

File: component/core/src/main/java/org/exoplatform/social/core/storage/api/SpaceStorage.java [hunk 2]
```diff
@@ -79,6 +79,24 @@
...
+
+  /**
+   * Gets the list of users who ignored the space
+   *
+   * @param spaceId the id of the space
+   * @param userId the id of the user
+   * @return true if user has ignored the space from suggested list
+   */
+
+  boolean isSpaceIgnored(String spaceId, String userId);
```
Comment (thomasdelhomenie) [inline component/core/src/main/java/org/exoplatform/social/core/storage/api/SpaceStorage.java]: The method signature does not match with the comments.

File: component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/dao/SpaceMemberDAO.java [hunk 3]
```diff
@@ -27,5 +27,9 @@
 
     SpaceMemberEntity getMember(String remoteId, Long spaceId);
 
+    SpaceMemberEntity getSpaceMemberShip(String remoteId, Long spaceId);
```
Comment (thomasdelhomenie) [inline component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/dao/SpaceMemberDAO.java]: We add many news methods in interfaces, it breaks the API. We should first check if we can reduce the number of new/updated methods in these interfaces. And if it is really necessary we should use default method to make the upgrade seamless.

File: component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/RDBMSSpaceStorageImpl.java (hunk 1 above)
Comment (boubaker) [inline component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/RDBMSSpaceStorageImpl.java]: +1

File: component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/dao/SpaceMemberDAO.java (hunk 3 above)
Comment (boubaker) [inline component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/dao/SpaceMemberDAO.java]: +1 we should add a method that gets the user membership in a space using three parameters:
- User id
- Space id
- Membership status
This should be a composite id with of course the test on uniqueness), but unfortunately, the model didn't use this from the beginning.

So @AtiaZ please combine getMember & getSpaceMemberShip to use a single method where:
- if the passed "status" parameter is null => get the associated SpaceMemberShip to selected user + space
- if the passed "status" different to null, use a different NamedQuery to select the SpaceMembership.

File: component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/dao/jpa/SpaceMemberDAOImpl.java [hunk 4]
```diff
@@ -38,6 +38,31 @@ public void deleteBySpace(SpaceEntity entity) {
   }
 
   @Override
+  public SpaceMemberEntity getSpaceMemberShip(String remoteId, Long spaceId, SpaceMemberEntity.Status status) {
```
Comment (boubaker) [inline component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/dao/jpa/SpaceMemberDAOImpl.java]: Test if parameters are null of empty, if not valide, please throw IllegalArgumenException.

File: component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/dao/jpa/SpaceMemberDAOImpl.java [hunk 5]
```diff
@@ -38,6 +38,31 @@ public void deleteBySpace(SpaceEntity entity) {
...
+    TypedQuery<SpaceMemberEntity> query = getEntityManager().createNamedQuery("SpaceMember.getMember", SpaceMemberEntity.class);
+    query.setParameter("userId", remoteId);
+    query.setParameter("spaceId", spaceId);
+    query.setParameter("status", status);
+    try {
+      return query.getSingleResult();
```
Comment (boubaker) [inline component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/dao/jpa/SpaceMemberDAOImpl.java]: We are not sure that we will have at most a unique result (No 'unique' condition on those three fields), so IMO we should manage the fact that we get more than one result and return the first one +logging an error saying that we have found more than one result and the data is not coherent. Or you can add a unique condition in Database model (which could have an impact in this stage of product version).

File: component/service/src/main/java/org/exoplatform/social/rest/api/EntityBuilder.java [hunk 6]
```diff
@@ -388,6 +388,28 @@ public static RelationshipEntity buildEntityRelationship(Relationship relationsh
     return infos;
   }
```
Comment (boubaker) [inline component/service/src/main/java/org/exoplatform/social/rest/api/EntityBuilder.java]: Could you add a documentation for this method plz

File: component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/RDBMSSpaceStorageImpl.java (hunk 1 above)
Comment (AtiaZ) [inline component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/RDBMSSpaceStorageImpl.java]: method renamed to "ignoreSpace"

File: component/service/src/main/java/org/exoplatform/social/rest/api/EntityBuilder.java (hunk 6 above)
Comment (AtiaZ) [inline component/service/src/main/java/org/exoplatform/social/rest/api/EntityBuilder.java]: done

**Comment by thomasdelhomenie** [comment]
@AtiaZ Due to recent changes in social code, this PR does not compile anymore. Please update it.

**Comment by AtiaZ** [comment]
PR updated.

File: component/core/src/main/java/org/exoplatform/social/core/space/spi/SpaceService.java [hunk 7]
```diff
@@ -969,6 +969,25 @@
...
 
   /**
+   * Checks if a user is in the list of users who have ignored the space.
+   *
+   * @param space The existing space.
+   * @param userId The remote user Id.
+   * @return TRUE if the space is an ignored space. Otherwise, it
+   *         is FALSE.
+   */
+   boolean isIgnored(Space space, String userId);
```
Comment (thomasdelhomenie) [inline component/core/src/main/java/org/exoplatform/social/core/space/spi/SpaceService.java]: These addition breaks the API. Please use a default method (which does nothing) instead.

Comment (AtiaZ) [inline component/core/src/main/java/org/exoplatform/social/core/space/spi/SpaceService.java]: done

File: component/core/src/main/java/org/exoplatform/social/core/storage/cache/CachedSpaceStorage.java [hunk 8]
```diff
@@ -355,6 +355,24 @@ public void deleteSpace(final String id) throws SpaceStorageException {
 
   }
 
+  @Override
+  public void ignoreSpace(String spaceId, String userId) {
+    storage.ignoreSpace(spaceId, userId);
+    Space space = storage.getSpaceById(spaceId);
```
Comment (boubaker) [inline component/core/src/main/java/org/exoplatform/social/core/storage/cache/CachedSpaceStorage.java]: Why getting space before clearing the cache ?

File: component/core/src/main/java/org/exoplatform/social/core/storage/impl/SpaceStorageImpl.java [hunk 9]
```diff
@@ -964,6 +963,18 @@ public void deleteSpace(String id) throws SpaceStorageException {
     );
   }
 
+  //This method is added to fix suggestion issue with RDBMS only
```
Comment (boubaker) [inline component/core/src/main/java/org/exoplatform/social/core/storage/impl/SpaceStorageImpl.java]: Use "API Documentation comment" or @inheritedDoc (if there is a doc comment defined in Interface) plz:
/**
 * 
 */

Comment (AtiaZ) [inline component/core/src/main/java/org/exoplatform/social/core/storage/impl/SpaceStorageImpl.java]: done. @inheritedDoc is used.

File: component/core/src/main/java/org/exoplatform/social/core/storage/cache/CachedSpaceStorage.java (hunk 8 above)
Comment (AtiaZ) [inline component/core/src/main/java/org/exoplatform/social/core/storage/cache/CachedSpaceStorage.java]: to use it when calling the cleanRef(space) method.

Comment (boubaker) [inline component/core/src/main/java/org/exoplatform/social/core/storage/cache/CachedSpaceStorage.java]: You can get from:
SpaceData spaceData = exoSpaceCache.remove(new SpaceKey(space.getId()));
Space space = spaceData.build();

Comment (AtiaZ) [inline component/core/src/main/java/org/exoplatform/social/core/storage/cache/CachedSpaceStorage.java]: done

File: component/core/src/main/java/org/exoplatform/social/core/storage/cache/CachedSpaceStorage.java [hunk 10]
```diff
@@ -355,6 +355,24 @@ public void deleteSpace(final String id) throws SpaceStorageException {
 
   }
 
+  @Override
+  public void ignoreSpace(String spaceId, String userId) {
+    storage.ignoreSpace(spaceId, userId);
+    exoSpaceSimpleCache.remove(new SpaceKey(spaceId));
+    SpaceData spaceData = exoSpaceCache.remove(new SpaceKey(spaceId));
+    Space space = spaceData.build();
```
Comment (boubaker) [inline component/core/src/main/java/org/exoplatform/social/core/storage/cache/CachedSpaceStorage.java]: Should test if spaceData is null

File: component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/dao/SpaceMemberDAO.java (hunk 3 above)
Comment (boubaker) [inline component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/dao/SpaceMemberDAO.java]: Could we combine both methods as suggested in previous comment?

File: component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/dao/jpa/SpaceMemberDAOImpl.java (hunk 5 above)
Comment (boubaker) [inline component/core-jpa/src/main/java/org/exoplatform/social/core/jpa/storage/dao/jpa/SpaceMemberDAOImpl.java]: ok, done apparently

File: component/core/src/main/java/org/exoplatform/social/core/storage/cache/CachedSpaceStorage.java (hunk 10 above)
Comment (AtiaZ) [inline component/core/src/main/java/org/exoplatform/social/core/storage/cache/CachedSpaceStorage.java]: done

**Comment by boubaker** [comment]
Remains on remark:
https://github.com/exodev/social/pull/124#discussion_r98947449

**Comment by AtiaZ** [comment]
PR updated

**Comment by AtiaZ** [comment]
Merged in develop.

**Pull request closed without merging**